lib/lookup/src/lookup_v2/jit.rs

```diff
@@ -0,0 +1,216 @@
+                            Some(BorrowedSegment::Field(&self.path[start..]))
+                        }
+                        JitState::Quote { .. } => Some(BorrowedSegment::Invalid),
+                        JitState::End => None,
+                    };
+                    self.state = JitState::End;
+                    return result;
+                }
+                Some((index, c)) => {
+                    let (result, state) = match self.state {
+                        JitState::Start | JitState::Continue => match c {
+                            '.' => (None, JitState::Dot),
+                            'A'..='Z' | 'a'..='z' | '_' | '0'..='9' => {
+                                (None, JitState::Field { start: index })
+                            }
+                            '[' => (None, JitState::IndexStart),
+                            '\"' => (None, JitState::Quote { start: index + 1 }),
+                            _ => (Some(Some(BorrowedSegment::Invalid)), JitState::End),
+                        },
+                        JitState::Dot => match c {
+                            'A'..='Z' | 'a'..='z' | '_' | '0'..='9' => {
+                                (None, JitState::Field { start: index })
+                            }
+                            '\"' => (None, JitState::Quote { start: index + 1 }),
+                            _ => (Some(Some(BorrowedSegment::Invalid)), JitState::End),
+                        },
+                        JitState::Field { start } => match c {
+                            'A'..='Z' | 'a'..='z' | '_' | '0'..='9' => {
+                                (None, JitState::Field { start })
+                            }
+                            '.' => (
+                                Some(Some(BorrowedSegment::Field(&self.path[start..index]))),
+                                JitState::Dot,
+                            ),
+                            '[' => (
+                                Some(Some(BorrowedSegment::Field(&self.path[start..index]))),
+                                JitState::IndexStart,
+                            ),
+                            _ => (Some(Some(BorrowedSegment::Invalid)), JitState::End),
+                        },
+                        JitState::Quote { start } => match c {
+                            '\"' => (
+                                Some(Some(BorrowedSegment::Field(&self.path[start..index]))),
+                                JitState::Continue,
+                            ),
+                            _ => (None, JitState::Quote { start }),
+                        },
+                        JitState::IndexStart => match c {
+                            '0'..='9' => (
+                                None,
+                                JitState::Index {
+                                    value: c as usize - '0' as usize,
+                                },
+                            ),
```

Comment:
This wasn't pointed out in the PR description, butt it's worth mentioning that for this lookup code to be supported in VRL, we also need to add support for negative indexing.

Reply:
I added it to https://github.com/vectordotdev/vector/issues/7070